c-deps/libroach/db.cc

```diff
@@ -527,6 +527,40 @@ DBStatus DBEnvLinkFile(DBEngine* db, DBSlice oldname, DBSlice newname) {
   return db->EnvLinkFile(oldname, newname);
 }
 
+DBIterState DBCheckForAddSSTableKeyCollisions(DBIterator* existingIter, DBIterator* sstIter) {
+  DBIterState state = {};
+  while (existingIter->rep->Valid() && sstIter->rep->Valid()) {
+    rocksdb::Slice sstKey;
+    rocksdb::Slice existingKey;
+    int64_t wall_time = 0;
+    int32_t logical_time = 0;
+    if (!DecodeKey(sstIter->rep->key(), &sstKey, &wall_time, &logical_time) ||
+        !DecodeKey(existingIter->rep->key(), &existingKey, &wall_time, &logical_time)) {
+          state.valid = false;
+          state.status = FmtStatus("unable to decode key");
+          return state;
+    }
+
+    int compare = kComparator.Compare(existingKey, sstKey);
+    if (compare == 0) {
+      // If the colliding key has been deleted in rocksdb then it is not
```

Comment:
to clarify, this is if it is a tombstone from an MVCC delete, right?

speaking of MVCC, are we missing anything here? seems like we might need to be handling intents or inline or something? cc ... @nvanbenschoten maybe?

Reply:
I'm not really sure what "collision" means in this context, so I agree with your point above. Are we defining collision as keys that overlap regardless of timestamps or exact (key, timestamp) collisions? A solid comment on this method would help a lot in my ability to determine whether this is implementing its desired behavior.

Reply:
I have added a more detailed comment in libroach.h, and clarified that a collision is an overlap of the key slice regardless of timestamps.

Reply:
The way this currently behaves is that it will return invalid if an overlapping key in `existingIter` contains anything other than one or more committed deletion tombstones. I think that gives you the behavior that you want, but it's pretty conservative. It will return invalid even for cases that I think you want to support like a collision with a deletion tombstone over a previously live value. Do we want to extend it to support those cases?

I'd also suggest you pull out the case where `logical_time == kZeroTimestamp` into an explicit case that describes our desired behavior for inlined values and intents.